### PR TITLE
Send message when spellbot does not have permission to manage reactions

### LINKED ISSUE #120: Issue warning when permissions are wrong
**Describe the solution you'd like**

When `safe_remove_reaction` fails due to permissions issues we can issue a warning in the channel to let the server admins know that the bot has incorrect permissions so they can potentially take action to resolve the issue.

We should also add a way to silence this warning so that admins who don't want to give SpellBot those permissions can make this warning go away.

### PR DESCRIPTION
**Pull request by crookedneighbor**
**Description**

Has spellbot message when it cannot adjust reactions because it lacks the proper permissions.

* Definitely not finished, since I didn't implement a way for admins to silence this message
* I chose to put the reactions code into their own file to make unit testing easier. If you'd prefer to consider these methods as private functions in the main test file, does that mean you want each place that calls them to be tested for this permissions message?
* Can't figure out why the meta test for strings doesn't work for this. Probably something to do with it being in a separate file?
* If you want to start pulling some of this code into smaller, more manageable files, is there a more programatic way to order the test files to ensure the Meta class gets invoked last, but not have to list every single test file?

- First step towards resolving #120 

**Checklist**

- [x] Add tests for these changes
- [x] Test coverage is not decreased
- [x] Entire test suite passes

### CODE REVIEW

**Comment by lexicalunit** [comment]
I think the meta test for strings is failing for 2 reasons:

1. It's only in the `client` fixture that we run `monkeypatch.setattr(spellbot, "s", S_SPY)` and none of the tests in `tests/test_reactions.py` use the `client` fixture.
2. Even if we add the `client` fixture to those tests it still won't work because we're actually monkey patching the `s()` function within the `spellbot` module. But that leaves the `s()` that's within the `reactions` module untouched.

So I was able to get the test to pass by adding the `client` fixture to the reactions tests and adding `monkeypatch.setattr(spellbot.reactions, "s", S_SPY)` to the client fixture. There's probably a nice way to refactor this code that it works and is more elegant.

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 - Adds meta theme-color and meta og:site_name to front page.
 - Adds a note about Administrator permissions to the Commands for Admins section of the front page.
+- Spellbot will message in the channel when they do not have permission to adjust reactions
```
Comment (lexicalunit) [inline CHANGELOG.md:18]:
```suggestion
- Spellbot will message in the channel when it does not have permission to adjust reactions.
```

File: src/spellbot/reactions.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import discord
+
+from spellbot.assets import s
+
+logger = logging.getLogger(__name__)
+
+
+async def safe_remove_reaction(
+    message: discord.Message, emoji: str, user: discord.User
+) -> None:  # pragma: no cover
```
Comment (lexicalunit) [inline src/spellbot/reactions.py:12]: We probably want to remove `pragma: no cover` here since we now have tests that cover this code 🎉

**Comment by lexicalunit** [comment]
>  is there a more programatic way to order the test files to ensure the Meta class gets invoked last, but not have to list every single test file?
For now I don't mind having to specify the order manually. If it gets to be annoying down the road we can probably just refactor `pytest_collection_modifyitems()` such that it just puts `TestCodebase` and `TestMeta` at the end and leaves everything else alone. But again, for now I don't mind is so much. 🤷‍♀️

**Comment by lexicalunit** [comment]
> Definitely not finished, since I didn't implement a way for admins to silence this message

We could break that off into a separate follow-on PR after we get this one merged. In the interest of keeping PRs to a reasonable size.

File: src/spellbot/assets/strings.yaml [hunk 3]
```diff
@@ -66,6 +66,8 @@ not_a_command: Sorry $reply, there is no "$request" command.
 not_admin: $reply, you do not have admin permissions to run that command.
 play_found: 'I found a game for you, $reply. You have been signed up for it! Go to
   game post: $link'
+reaction_permissions_required: I do not have permission to adjust reactions. A server
```
Comment (lexicalunit) [inline src/spellbot/assets/strings.yaml:69]: Nice 👌

**Comment by crookedneighbor** [comment]
Thanks for the help!

**Comment by crookedneighbor** [comment]
I believe the coverage has gone down because we're not testing the code path where the logger gets called. Wasn't sure how you wanted to do that though.

**Comment by lexicalunit** [comment]
> I believe the coverage has gone down because we're not testing the code path where the logger gets called. Wasn't sure how you wanted to do that though.

Here's an additional test for `safe_clear_reactions()` that gives us 100% coverage for that function:

```python
    async def test_clear_reactions_when_http_error_is_thrown(self, caplog):
        message = MockDiscordMessage()
        fake_response = Mock()
        fake_response.status = 400
        with patch.object(message, "clear_reactions", wraps=message.clear_reactions) as m:
            m.side_effect = discord.errors.HTTPException(fake_response, "BAD REQUEST")
            with patch.object(message.channel, "send", wraps=message.channel.send):
                await safe_clear_reactions(cast(discord.Message, message))

                assert "warning: discord: could not clear reactions" in caplog.text
                assert "BAD REQUEST" in caplog.text
```

I am doing the `cast(discord.Message, message)` to make Pylance shut up about the type mismatch there. But the real meat of the test is that we can use the `caplog` fixture to make sure that the log contains the things we want. While working on this though I found a bug!

In `reactions.py` I have this code:

```python
logger.exception("warning: discord: could not clear reactions", e)
```

But that is actually wrong. It needs to be:

```python
logger.exception("warning: discord: could not clear reactions: %s", e)
```

If you want to fix up all the `logger.exception()` calls in this PR I would be very grateful!

File: tests/test_reactions.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                message.channel, "send", wraps=message.channel.send
+            ) as fake_channel_send:
+                await safe_remove_reaction(message, "emoji", FRIEND)
+
+                fake_channel_send.assert_called_with(
+                    (
+                        "I do not have permission to adjust "
+                        "reactions. A server admin will need to adjust my permissions."
+                    )
+                )
```
Comment (lexicalunit) [inline tests/test_reactions.py:45]: We can actually remove these extra parens here:

```suggestion
                fake_channel_send.assert_called_with(
                    "I do not have permission to adjust "
                    "reactions. A server admin will need to adjust my permissions."
                )
```

File: tests/test_reactions.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                message.channel, "send", wraps=message.channel.send
+            ) as fake_channel_send:
+                await safe_clear_reactions(message)
+
+                fake_channel_send.assert_called_with(
+                    (
+                        "I do not have permission to adjust "
+                        "reactions. A server admin will need to adjust my permissions."
+                    )
+                )
```
Comment (lexicalunit) [inline tests/test_reactions.py:78]: And here:

```suggestion
                fake_channel_send.assert_called_with(
                    "I do not have permission to adjust "
                    "reactions. A server admin will need to adjust my permissions."
                )
```

**Comment by crookedneighbor** [comment]
Sorry I didn’t get a chance to fix up the logging yesterday. Your changes look :+1: to me.

**Comment by lexicalunit** [comment]
> Sorry I didn’t get a chance to fix up the logging yesterday. Your changes look 👍 to me.

I was mostly testing out https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork to see if it actually worked. Pretty neat.

**Pull request merged**